marklogic-data-hub/src/main/java/com/marklogic/hub/impl/HubProjectImpl.java

```diff
@@ -542,6 +555,67 @@ public void upgradeProject() throws IOException {
+        File finalDbFile = getUserConfigDir().resolve("database-fields").resolve("final-database.xml").toFile();
+        try {
+            FileInputStream fileInputStream = new FileInputStream(finalDbFile);
+            DocumentBuilderFactory documentBuilderFactory = DocumentBuilderFactory.newInstance();
+            DocumentBuilder documentBuilder = documentBuilderFactory.newDocumentBuilder();
+
+            Document document = documentBuilder.parse(fileInputStream);
+
+            XPathFactory xPathFactory = XPathFactory.newInstance();
+            XPath xPath = xPathFactory.newXPath();
+
+            XPathExpression expr = xPath.compile("//range-path-index/*[local-name()='path-expression']/text()='//actionDetails/*/uris'");
+
+            Boolean isNodePresent = Boolean.parseBoolean(expr.evaluate(document));
+
+            if(!isNodePresent) {
+                Node node = (Node) xPath
+                    .evaluate("//*[local-name()='range-path-indexes']", document.getDocumentElement(), XPathConstants.NODE);
+
+                Node newNode = node.appendChild(document.createElement("range-path-index"));
+
+                Element scalarType = document.createElement("scalar-type");
+                scalarType.setTextContent("string");
+
+                Element collation = document.createElement("scalar-type");
+                collation.setTextContent("http://marklogic.com/collation/");
+
+                Element pathExpression = document.createElement("path-expression");
+                pathExpression.setTextContent("//actionDetails/*/uris");
+
+                Element rangeValuePosition = document.createElement("range-value-positions");
+                rangeValuePosition.setTextContent("true");
+
+                Element invalidValues = document.createElement("invalid-values");
+                invalidValues.setTextContent("reject");
+
+                newNode.appendChild(scalarType);
+                newNode.appendChild(collation);
+                newNode.appendChild(pathExpression);
+                newNode.appendChild(rangeValuePosition);
+                newNode.appendChild(invalidValues);
+
+                TransformerFactory tf = TransformerFactory.newInstance();
+                Transformer transformer = tf.newTransformer();
+                transformer.setOutputProperty(OutputKeys.OMIT_XML_DECLARATION, "yes");
+                transformer.setOutputProperty(OutputKeys.INDENT, "yes");
+                transformer.setOutputProperty(OutputKeys.ENCODING, "UTF-8");
+                transformer.setOutputProperty("{http://xml.apache.org/xslt}indent-amount", "2");
+
+                transformer.transform(new DOMSource(document),
+                    new StreamResult(new OutputStreamWriter(new FileOutputStream(finalDbFile), "UTF-8")));
+            }
+        }
+        catch (Exception e) {
+            throw new DataHubProjectException("Unable to upgrade the project: Failed to add index to final database", e);
```

Comment:
I think this is okay. I was initially worried that by throwing the exception, the user would think that hubUpgrade failed. But in reality, almost all of it succeeded except the last step. I think the error message could do a better job in explaining what failed - e.g. "Error while upgrading project; was not able to add //actionDetails/*/uris path range index to final-database.xml file; cause: " + ex.getMessage(). 

@srinathgit  Thoughts on throwing the error, or just logging it?

Reply:
I like throwing the exception and will change the error message to the one you had suggested